lib/krikri/harvesters/async_uri_getter.rb

```diff
@@ -0,0 +1,74 @@
+require 'net/http'
+require 'thread'
+require 'uri'
+
+# Get URIs asynchronously
+class AsyncUriGetter
+  def initialize(opts: {})
+    super()
+    @follow_redirects = follow_redirects(opts: opts)
+  end
+
+  def add_request(uri: '', headers: {}, opts: {})
+    raise 'uri must be a URI' unless uri.is_a?(URI)
+    Request.new(uri, headers, follow_redirects: follow_redirects(opts))
+  end
+
+  Request = Struct.new(:uri, :headers, :opts) do
+    def initialize(*)
+      super
+      @follow_redirects = opts[:follow_redirects]
+      @request_thread = start_request
+    end
+
+    def join
+      @request_thread.join
+    end
+
+    def with_response
+      yield(@request_thread.value)
+    end
+
+    private
+
+    def start_request
+      Thread.new do
+        response = nil
+        uri_to_try = uri
+        10.times do
+          http = Net::HTTP.new(uri_to_try.host, uri_to_try.port)
```

Comment:
:+1: to this. We also use Faraday in the OAI harvester: https://github.com/dpla/KriKri/blob/develop/lib/krikri/harvesters/oai_harvester.rb#L42

It would be nice to be able to have its benefits application-wide.